include/nbl/video/utilities/CArithmeticOps.h

```diff
@@ -0,0 +1,286 @@
+		struct SchedulerParameters : nbl::hlsl::scan::DefaultSchedulerParameters_t  // this struct and its methods are also available in HLSL
+		{
+			SchedulerParameters()
+			{
+				std::fill_n(cumulativeWorkgroupCount,Parameters::MaxLevels,0u);
+				std::fill_n(workgroupFinishFlagsOffset, Parameters::MaxLevels, 0u);
+				std::fill_n(lastWorkgroupSetCountForLevel,Parameters::MaxLevels,0u);
+			}
+            
+            // given the number of elements and workgroup size, figure out how many atomics we need
+            // also account for the fact that we will want to use the same scratch buffer both for the
+            // scheduler's atomics and the aux data storage
+			SchedulerParameters(Parameters& outScanParams, const uint32_t _elementCount, const uint32_t workgroupSize) : SchedulerParameters()
+			{
+				outScanParams = Parameters(_elementCount,workgroupSize);
+				const auto topLevel = outScanParams.topLevel;
+
+				std::copy_n(outScanParams.lastElement+1u,topLevel,cumulativeWorkgroupCount);
+				for (auto i=0u; i<=topLevel; i++)
+					cumulativeWorkgroupCount[i] += 1u;
+				std::reverse_copy(cumulativeWorkgroupCount,cumulativeWorkgroupCount+topLevel,cumulativeWorkgroupCount+topLevel+1u);
+
+				for (auto i = 0u; i <= topLevel; i++) {
+					workgroupFinishFlagsOffset[i] = ((cumulativeWorkgroupCount[i] - 1u) >> hlsl::findMSB(workgroupSize - 1u)) + 1;
+					lastWorkgroupSetCountForLevel[i] = cumulativeWorkgroupCount[i] & (workgroupSize - 1u);
+				}
+
+				const auto wgFinishedFlagsSize = std::accumulate(workgroupFinishFlagsOffset, workgroupFinishFlagsOffset + Parameters::MaxLevels, 0u);
+				std::exclusive_scan(workgroupFinishFlagsOffset, workgroupFinishFlagsOffset + Parameters::MaxLevels, workgroupFinishFlagsOffset, 0u);
+				for (auto i=0u; i<sizeof(Parameters::temporaryStorageOffset)/sizeof(uint32_t); i++)
+					outScanParams.temporaryStorageOffset[i] += wgFinishedFlagsSize;
+				
+				std::inclusive_scan(cumulativeWorkgroupCount, cumulativeWorkgroupCount + Parameters::MaxLevels, cumulativeWorkgroupCount);
+			}
+		};
+        
+        // push constants of the default direct scan pipeline provide both aux memory offset params and scheduling params
+		struct DefaultPushConstants
+		{
+			Parameters scanParams;
+			SchedulerParameters schedulerParams;
+		};
+        
+		struct DispatchInfo
+		{
+			DispatchInfo() : wg_count(0u)
+			{
+			}
+            
+            // in case we scan very few elements, you don't want to launch workgroups that wont do anything
+			DispatchInfo(const IPhysicalDevice::SLimits& limits, const uint32_t elementCount, const uint32_t workgroupSize)
+			{
+				constexpr auto workgroupSpinningProtection = 4u; // to prevent first workgroup starving/idling on level 1 after finishing level 0 early
+				wg_count = limits.computeOptimalPersistentWorkgroupDispatchSize(elementCount,workgroupSize,workgroupSpinningProtection);
+			}
+
+			uint32_t wg_count;
+		};
```

Comment:
move everything to HLSL, you can also have a look at PR #643 where the `computeOptimalPErsistentWorkgroupDispatchSize` may move to `device_capability_traits`